packages/compass-aggregations/src/modules/pipeline-builder/pipeline-mode.ts

```diff
@@ -26,9 +29,27 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
   return state;
 };
 
-export const changePipelineMode = (mode: PipelineMode): PipelineModeToggledAction => ({
-  type: ActionTypes.PipelineModeToggled,
-  mode,
-});
+export const changePipelineMode = (
+  newMode: PipelineMode
+): PipelineBuilderThunkAction<void, PipelineModeToggledAction> => {
+  return (dispatch, _getState, { pipelineBuilder }) => {
+    // Sync the PipelineBuilder
+    if (newMode === 'as-text') {
+      pipelineBuilder.stagesToSource();
+      dispatch(changeEditorValue(
+        pipelineBuilder.getPipelineStringFromStages()
+      ));
+    } else {
+      pipelineBuilder.sourceToStages();
+      dispatch(changeStages(pipelineBuilder.stages));
+    }
+
+    // Toggle the view (and state)
+    dispatch({
+      type: ActionTypes.PipelineModeToggled,
+      mode: newMode,
+    });
```

Comment:
I'd maybe suggest making this one event instead of three separate ones, something like:

```ts
if (newMode === 'as-text') {
  pipelineBuilder.stagesToSource();
} else {
  pipelineBuilder.sourceToStages();
}
dispatch({
  type: ActionTypes.PipelineModeToggled,
  source: pipelineBuilder.source,
  stages: pipelineBuilder.stages,
});
```

I think that way the UI flow is clearer in the reducers because you are avoiding actions that are too generic and lack context, this aligns with [redux style guide](https://redux.js.org/style-guide/#model-actions-as-events-not-setters) too